patchwork-registries/src/main/java/net/minecraftforge/registries/RegistryManager.java

```diff
@@ -43,28 +53,121 @@ public String getName() {
 
 	@SuppressWarnings("unchecked")
 	public <V extends IForgeRegistryEntry<V>> Class<V> getSuperType(Identifier key) {
-		return (Class<V>) RegistryClassMapping.getClass(key);
+		return (Class<V>) this.superTypes.inverse().get(key);
 	}
 
 	@SuppressWarnings("unchecked")
 	public <V extends IForgeRegistryEntry<V>> ForgeRegistry<V> getRegistry(Identifier key) {
-		return (ForgeRegistry<V>) this.registries.get(key);
+		Registry<?> vanillaRegistry = Registry.REGISTRIES.get(key);
+
+		if (vanillaRegistry instanceof ForgeRegistryProvider) {
```

Comment:
Alternatively, it might be a good idea to have a vanilla registry instance -> ForgeRegistry map here, instead of attaching it as a value of the registry instance with a mixin? Not entirely sure on this issue though.

Reply:
In the vanilla registry record, all forge modded registries are instance of ExtendedVanillaRegistry (super class of MutableRegistry), and they hold the reference to the its ForgeRegistry. The interface ForgeRegistryProvider is used for both forge and vanilla registries for retrieving ForgeRegistry reference.

Having a vanilla registry->ForgeRegistry map would work, but I still prefer to not maintaining a separate map.

Reply:
I also have a concern, currently vanilla and fabric mods can access registries added by forge, but not the other way around. How should we handle the registries created by fabric mods? Should we make them visible to forge mods?

Reply:
I cant imagin any of those cases... So our current code should be fine